expr/expr.go

```diff
@@ -24,209 +25,237 @@ var ErrNoSuchFunction = errors.New("no such function")
 var ErrNotContainer = errors.New("cannot apply in to a non-container")
 var ErrBadCast = errors.New("bad cast")
 
-type NativeEvaluator func(*zng.Record) (zngnative.Value, error)
-
-// CompileExpr tries to compile the given Expression into a function
-// that evalutes the expression against a provided Record.  Returns an
+// CompileExpr compiles the given Expression into an object
+// that evalutes the expression against a provided Record.  It returns an
 // error if compilation fails for any reason.
 //
-// This is currently not particularly optimized -- it creates a bunch
-// of closures and every evaluation involves some allocations.
-// Eventually, we could optimize this by compiling a particular
-// Expression for a particular TypeRecord into a series of byte codes
-// that could be implemented by a simple stack-based evaluator much
-// more efficiently.  ZNG unions are a challenge for this approach, but
-// we could fail back to the "slow path" implemented here if an
-// expression ever touches a union.
-func CompileExpr(node ast.Expression) (ExpressionEvaluator, error) {
-	ne, err := compileNative(node)
-	if err != nil {
-		return nil, err
-	}
-
-	return func(rec *zng.Record) (zng.Value, error) {
-		nv, err := ne(rec)
-		if err != nil {
-			return zng.Value{}, err
-		}
-
-		return nv.ToZngValue()
-	}, nil
+// This is the "intepreted slow path" of the analytics engine.
+// It is completely adaptive to the dynamic type system of zng so there
+// are a lot of types checks and run-tme decisions made according to type.
+//
+// Eventually, we will optimize this by adding a "fast path" that dynamically
+// generates byte codes (which an in turn be JIT assembled into machine code)
+// for each zng TypeRecord encountered.  Once you know the type record,
+// you can generate code using strong typing just as an OLAP system does
+// due to its schemas defined up-front in its relational tables.  Here,
+// each record type is like a schema and as we encounter them, we can compile
+// optimized code for the now-static types within that record type.
+//
+// The keep flag, if true, says to return zng.Value that are safe to stash
```

Comment:
Looking at the uses in the PR, it looks like nearly all calls CompileExpr uses keep = false. What do you think about either of these different approaches:
opt 1: Make all calls to CompileExpr/Eval act like "keep = false". Change the Evaluator interface to use a method name like "SharedValEval" (or some other name that clarifies a shared result is being returned). Add a convenience copy utility that takes a value and returns a copy of it for the uses where we'd otherwise use 'keep = true'.
opt 2: Make zng.Value similar to zng.Record: it has a 'nonvolatile' flag, and `Keep` and `CopyBody` functions. Remove the keep parameter from CompileExpr, and have all Eval's return volatile zng.Value.

I like option 2 for the similarity with zng.Record handling, so devs have a common set of idioms to use.

Reply:
I'm averse to option 2 because it adds overhead to zng.Value which is passed around extensively in the inner loop.  How about option 3?  Change Eval to always return a potentially volatile expression but introduce `expr.Keep(zng.Value)` to make copies of the byte slices in a Value as package expr knows which thingy-s are volatile.

Reply:
> I'm averse to option 2 because it adds overhead to zng.Value which is passed around extensively in the inner loop. How about option 3? Change Eval to always return a potentially volatile expression but introduce `expr.Keep(zng.Value)` to make copies of the byte slices in a Value as package expr knows which thingy-s are volatile. Basically option 1 but with a new function rather than a method on Evaluator.

+1 - that's what's I was trying to describe as option 1; that sounds good to me.

Reply:
This was a really good idea.  It forced me to consider exactly which calls sites needed to make a copy of the zng.Value.  Strings and IPs will be allocated twice by the logic here, but rather than check if the copied type s one of these, we'll just copy everything (and fortunately these copies don't happen in the inner loop) and I made issue #1412 to fix the string/IP issue.

Reply:
Just pushed the change for this...